File: src/main/java/io/lsdconsulting/lsd/distributed/interceptor/captor/http/derive/PathDeriver.java
package io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive;

import org.springframework.http.HttpRequest;

public class PathDeriver {
    private static final String EXTRACT_PATH = "https?://.*?(/.*)";

    public String derivePathFrom(final String url) {
        String path = url.replaceAll(EXTRACT_PATH, "$1");
        if (path.equals(url)) {
            return "";
        }
        return path;
    }

    public String derivePathFrom(final HttpRequest request) {
        return request.getURI().getPath() + (request.getURI().getQuery() != null ? "?" + request.getURI().getQuery() : "");
    }
}

File: src/main/java/io/lsdconsulting/lsd/distributed/interceptor/captor/http/derive/PathDeriver.kt
package io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive

import org.springframework.http.HttpRequest

open class PathDeriver {
    fun derivePathFrom(url: String): String {
        val path = url.replace(EXTRACT_PATH.toRegex(), "$1")
        return if (path == url) {
            ""
        } else path
    }

    fun derivePathFrom(request: HttpRequest): String =
        request.uri.path + if (request.uri.query != null) "?" + request.uri.query else ""

    companion object {
        private const val EXTRACT_PATH = "https?://.*?(/.*)"
    }
}
